devolutions-crypto-wayk/src/bastion/fle.rs

```diff
@@ -0,0 +1,203 @@
+use crate::bastion::Error;
+use block_padding::{Padding, Pkcs7};
+use chacha20poly1305::aead::{AeadInPlace, NewAead};
+use chacha20poly1305::{Key, XChaCha20Poly1305, XNonce};
+use uuid::Uuid;
+
+pub const AAD_MAGIC_NUMBER: [u8; 4] = [0x53, 0x0A, 0x0F, 0x0E];
+pub const SAFE_PAYLOAD_PREFIX: &str = "|~Safe~|";
+pub const DEFAULT_BLOCK_SIZE: usize = 32;
+
+/// Encrypts field-level content using default padding size.
+///
+/// See `encrypt_field_with_block_size` for details.
+pub fn encrypt_field(master_key_id: Uuid, key: &[u8], content: &[u8]) -> Result<String, Error> {
+    encrypt_field_with_block_size(master_key_id, key, content, DEFAULT_BLOCK_SIZE)
+}
+
+/// Encrypts field-level content using provided padding size.
+///
+/// - Output is the encrypted content in Bastion safe Field Value Format (ie: |~Safe~|k1AjQZYciYvlPmJMnHjbb…)
+///
+/// # Interals
+///
+/// - AAD is composed of a 4-byte magic number, a 4-byte reserved field and the master key UUID in big-endian binary format.
+/// - A 24-byte initialization vector (IV) is randomly generated.
+/// - Content is padded using PKCS#7 padding mode.
+/// - Encryption using XChaCha2020-Poly1035.
+/// - Output is a buffer such as `[magic number (4) | reserved (4) | UUID (16) | IV (24) | Ciphertext (variable) | Tag (16)]`
+///     encoded in Bastion safe Field Value Format.
+pub fn encrypt_field_with_block_size(
+    master_key_id: Uuid,
+    key: &[u8],
+    content: &[u8],
+    block_size: usize,
+) -> Result<String, Error> {
+    use rand::rngs::OsRng;
+    use rand::Fill;
+
+    let pad_leftover = block_size - (content.len() % block_size);
```

Comment:
Another way to see this is to do something like ((content.len() / BLOCK_SIZE) + 1) * BLOCK_SIZE, which will be the plaintext length, padding included, instead of trying to separate the two. You could also just pad the plaintext before allocating the buffer and use the length of the padded plaintext directly, this would be a bit more efficient.

Reply:
Indeed! Regarding allocation: If I pad before allocating, it will be two allocations instead of only one and it wouldn't eliminate padded size calculation. `Pcks7::pad` requires a buffer with at least `msg.len() + pad_len` as input and performs padding in-place. That being said, did I miss something?